### ISSUE TITLE
Expose Lexbor fragment parser

### ISSUE THREAD

**Issue by FeldrinH** [rushter/selectolax#74]
Currently any attempt to parse HTML fragments (such as for example `<div>One</div><div>Two</div>`) has to go through `LexborHTMLParser` which treats it as a HTML document and adds `<html>`, `<body>` and other unneeded tags.

Using `LexborHTMLParser(fragment).body.child`, which is shown in some examples gives a result that seems correct for common cases but for many edge cases is very different from the parsing algorithm used by browsers to parse HTML fragments.

Lexbor appears to have a special parser for HTML fragments (`lxb_html_document_parse_fragment`). It would be nice if this could be exposed in selectolax.

**Comment by JuroOravec** [comment]
Hey, I just came across this issue too!

**Context:**

In [django_components's](https://github.com/EmilStenstrom/django-components) components, I am given HTML which I do NOT know if it is full HTML document or a fragment.

If it is a fragment, like
```html
<div>
  <span>
    Hello
  </span>
</div>
```

Then `LexborHTMLParser` wraps this fragment HTML within `<html>`, `<body>`, `<head>` (extra lines added for clarity):

```html
<html>
  <head>
  </head>
  <body>

	<div>
	  <span>
	    Hello
	  </span>
	</div>

  </body>
</html>
```

I found a workaround to discern HTML documents from fragments, which is that I check if the entirety of the HTML string is within `<head>` or `<body>`. If so, then I'm dealing with a fragment, and I unwrap it.

However, there's a problematic edge case when the given HTML starts with tags that are permitted to be in the `<head>` (See explanation in https://github.com/lexbor/lexbor/issues/183#issuecomment-1611975340). E.g.:

```html
<link href="https://..." />
<div>
  <span>
    Hello
  </span>
</div>
```

Which gets turned into
```html
<html>
  <head>
    <link href="https://..." />
  </head>
  <body>

	<div>
	  <span>
	    Hello
	  </span>
	</div>

  </body>
</html>
```

---

Would the fragment parser work for this case, AKA would the fragment parser parse the HTML tags as is, without moving tags around nor adding extra tags?

And could I use the fragment parser to handle both full HTML documents and HTML fragments? Does the fragment parser keep e.g. `<!doctype>`, `<html>`, etc, unchanged?

@lexborisov, these are probably questions on you.

If so, then I can help implementing this onto selectolax.

**Comment by JuroOravec** [comment]
I played around this today. I tried different approaches, it compiled, but got seg faults at runtime. If someone more familiar with either this repo / Cython / C comes across this, let me know and I can share what I got.

One realization, though, is that the instantiation of the `Node` class requires the `HTMLParser` instance. So if I want to have a function like

```py
node = lexbor_fragment("<div>...</div>")
assert isinstance(node, LexborNode)
```

Then internally `lexbor_fragment` still needs to instantiate the `HTMLParser`, and then pass that to `Node`.

---

Based on the return types, I _think_ that what I want is [the value of `parser->root`](https://github.com/lexbor/lexbor/blob/72236d31da1a4c3d83c621446dcf4f70345fda05/source/lexbor/html/parser.c#L325).

I tried using something like [`lxb_html_parse_fragment`](https://github.com/lexbor/lexbor/blob/72236d31da1a4c3d83c621446dcf4f70345fda05/source/lexbor/html/parser.c#L148), but I think that I got seg faults because those functions also clean up after themselves.

So then I tried calling [`lxb_html_parse_fragment_chunk_begin`, `lxb_html_parse_fragment_chunk_process`,  `lxb_html_parse_fragment_chunk_end`](https://github.com/lexbor/lexbor/blob/72236d31da1a4c3d83c621446dcf4f70345fda05/source/lexbor/html/parser.c#L159) one-by-one, to see where the problem could be. But I didn't get farther.

**Comment by rushter** [comment]
@lexborisov Can you please give us some hints?

**Comment by lexborisov** [comment]
Hi @rushter @JuroOravec 

I'm not quite sure what's required of me? The task itself seems strange: to identify a full or a fragment of a document.
I'd say the task is unsolvable.

In HTML, the `<html>`, `<head>`, `< body>` tags are optional. They are created automatically during parsing.
In other words, for me the complete document is `<div>Oh my</div>`.

`<div>Oh my</div>` == `<html><head></head><body><div>Oh my</div></body></html>`.

In general, only a person can decide whether they have a fragment or a full document.

P.S.: I didn't quite understand about the segfault.

**Comment by JuroOravec** [comment]
Hey @lexborisov, basically I wonder if Lexbor / Modest has some function that parses HTML, but without inserting the `<html>`, `<head>`, `<body>` tags. 

That's why I was playing around with [`lxb_html_parse_fragment`](https://github.com/lexbor/lexbor/blob/72236d31da1a4c3d83c621446dcf4f70345fda05/source/lexbor/html/parser.c#L148). But I'm new to Lexbor codebase, so it's all just assumptions on my end.

If there is a way to parse HTML without adding the extra tags, then for Selectolax we could write a function in Cython, and use the corresponding Lexbor API.

Otherwise, we would have to do the data manipulation to remove the extra tags in Python layer.

**Comment by JuroOravec** [comment]
Btw, @rushter, what do you think about having these features defined in selectolax? I can have a go at it.

I also noticed the suggestion for `new_tag("th")` in #133. So new APIs could look e.g. like this:

```py
lexbor_parse(html: str | bytes) -> List[Union[LexborNode, LexborHTMLParser]]
lexbor_tag(tag_name: str) -> LexborNode

modest_parse(html: str | bytes) -> List[Union[Node, HTMLParser]]
modest_tag(tag_name: str) -> Node
```

- **`lexbor_tag` / `modest_tag`**
  - These would be just given a tag name, e.g. `h2`, and would create an empty Node. Same as
    ```py
    def lexbor_tag(tag_name: str):
        return lexbor_parse(f"<{tag_name}></{tag_name}>")[0]
    ```

- **`lexbor_parse` / `modest_parse`**
  - This would be similar to the implementation code I wrote at the end of [this comment](https://github.com/rushter/selectolax/issues/74#issuecomment-2404470344).
  - The idea is that while `HTMLParser` wraps the HTML in `<html>`, `<body>`, etc, `lexbor_parse` would return
    the nodes that corresponds to the given HTML. So it would return `HTMLParser` only if user gave the parser a proper HTML with `<html>` at the top-level. Otherwise it would return a list of `Node`.
  - To support multi-root HTML fragments, the results would be returned as a list.
  - Internally it would use `HTMLParser` to parser the HTML. Then it would somehow decide whether the given HTML was a fragment or not, and if so, would unpack `HTMLParser` into `Node`

  ```py
  def lexbor_parse(html: str | bytes):
      tree = LexborHTMLParser(html)
      if is_fragment(tree):
          return extract_fragment(tree, html)
      else:
          return [tree]
  ```

**Comment by rushter** [comment]
> Btw, @rushter, what do you think about having these features defined in selectolax? I can have a go at it.
> 
> I also noticed the suggestion for `new_tag("th")` in #133. So new APIs could look e.g. like this:
> 
> ```python
> lexbor_parse(html: str | bytes) -> List[Union[LexborNode, LexborHTMLParser]]
> lexbor_tag(tag_name: str) -> LexborNode
> 
> modest_parse(html: str | bytes) -> List[Union[Node, HTMLParser]]
> modest_tag(tag_name: str) -> Node
> ```
> 
> * **`lexbor_tag` / `modest_tag`**
>   
>   * These would be just given a tag name, e.g. `h2`, and would create an empty Node. Same as
>     ```python
>     def lexbor_tag(tag_name: str):
>         return lexbor_parse(f"<{tag_name}></{tag_name}>")[0]
>     ```
> * **`lexbor_parse` / `modest_parse`**
>   
>   * This would be similar to the implementation code I wrote at the end of [this comment](https://github.com/rushter/selectolax/issues/74#issuecomment-2404470344).
>   * The idea is that while `HTMLParser` wraps the HTML in `<html>`, `<body>`, etc, `lexbor_parse` would return
>     the nodes that corresponds to the given HTML. So it would return `HTMLParser` only if user gave the parser a proper HTML with `<html>` at the top-level. Otherwise it would return a list of `Node`.
>   * To support multi-root HTML fragments, the results would be returned as a list.
>   * Internally it would use `HTMLParser` to parser the HTML. Then it would somehow decide whether the given HTML was a fragment or not, and if so, would unpack `HTMLParser` into `Node`
>   
>   ```python
>   def lexbor_parse(html: str | bytes):
>       tree = LexborHTMLParser(html)
>       if is_fragment(tree):
>           return extract_fragment(tree, html)
>       else:
>           return [tree]
>   ```


What's the benefit of `lexbor_parse` instead of a more explicit option like `parse_fragment`?

**Comment by JuroOravec** [comment]
@rushter When I was writing previous message, I was thinking of having a function that would take some HTML, and would convert it to either `HTMLParser` or `Node`, depending on the structure of the given HTML (e.g. if it contains `<html>`, then treat it as  `HTMLParser`, otherwise as a list of Nodes.

And my line of thinking was that if there was such a function that returned either A or B, then it shouldn't be called `parse_fragment`, but just `parse`.

Buuuuuut, I had a further look at the codebase, and I realized that:

1. Initially I wasn't sure whether it would be possible to convert `HTMLParser` to `Node`. Because if we wanted to have `parse_fragment`, then IMO it should ALWAYS return instances of `Node`. So even if I pass in a full HTML document with `<html> ... </html>`, it should return Nodes. But I found that we can convert `HTMLParser` to `Node` via `HTMLParser.root`. So we can totally have `parse_fragment`

2. I also noticed that the API for `Node` and `HTMLParser` are almost identical (in Lexbor, it seems that `Node` is missing `clone()`. And I'm not sure what `LexborHTML.selector` does.

	The reason why I was thinking about the function that would return either `HTMLParser` or `Node` was because I was under impression that `HTMLParser` might have some API that would not be available if we had the same underlying HTML, but wrapped in `Node`.

	But IMO both `HTMLParser` and `Node` practically identical API, so IMO it's sufficient to have just `parse_fragment()`, and not the "smart" `parse()` 

So to summarize I'm in favour of `parse_fragment` over `parse`.

---

And to recap:
1. `lexbor/modest_tag()` as mentioned above
2. `lexbor/modest_parse_fragment()`, which always returns a list of Nodes. Even if it receives a full document with `<html></html>`
